Plugins: Add deprecation notice for /api/tsdb/query endpoint (#45238)
grafana/grafana: docs/sources/http_api/data_source.md

Proxies all calls to the actual data source.

## Query a data source

Queries a data source having a backend implementation.

`POST /api/ds/query`

> > **Note:** Most of Grafana's builtin data sources have a backend implementation.

Reviewer: ```suggestion
> **Note:** Grafana's built-in data sources usually have a backend implementation.
```

**Example request for the Test data source**:

```http
POST /api/ds/query HTTP/1.1
Accept: application/json
Content-Type: application/json

{
   "queries":[
      {
         "refId":"A",
         "scenarioId":"csv_metric_values",
         "datasource":{

>             "uid":"PD8C576611E62080A"

Reviewer 2: What about the type? Optional?

         },
         "format": "table"
         "maxDataPoints":1848,
         "intervalMs":200,
         "stringInput":"1,20,90,30,5,0",
      }
   ],
   "from":"now-5m",
   "to":"now"
}
```

> > **Note:** The `from`, `to`, and `queries` properties are required.

Reviewer: You do not need a note here. You can simply label the "from/to" and "queries" properties as required in the section below.

JSON Body schema:

> - **from/to** – Specifies the time range for the queries. Are both required and should be either epoch timestamps in milliseconds or relative using Grafana time units. For example, `now-5m`.

Reviewer: ```suggestion
- **from/to** – (Required) Specifies the time range for the queries. The timestamp can be either epoch timestamps in milliseconds or relative using Grafana time units. For example, `now-5m`.
```

> - **queries** – Specifies one or more queries. Must contain at least 1.

Reviewer: ```suggestion
- **queries** – (Required) Specifies one or more queries. Must contain at least 1.
```

> - **queries.refId** – Specifies an identifier of the query. Is optional and defaults to "A".

Reviewer: ```suggestion
- **queries.refId** – (Optional) Specifies an identifier of the query.  Defaults to "A".
```

> - **queries.datasource.uid** – Specifies the UID of data source to be queried. Each query in the request must have a unique `datasource`.

Reviewer 2: Not sure if this makes more sense. 

```suggestion
- **queries.datasource.uid** – Specifies the UID of data source to be queried. Each query in the request must have a `datasource`.
```

- **queries.format** – Specifies the format the data should be returned in. Valid options are `time_series` or `table` depending on the data source.
- **queries.maxDataPoints** - Species the maximum amount of data points that a dashboard panel can render. Is optional and defaults to 100.
- **queries.intervalMs** - Specifies the time series time interval in milliseconds. Is optional and defaults to 1000.

> In addition, each data source has its own specific properties that should be added in a request (for example **queries.stringInput** as shown in the request above). In order to understand how to form a query for a certain datasource, please use the Developer Tools in your browser of choice and inspect the HTTP requests being made to /api/ds/query.

Reviewer: ```suggestion
In addition, specific properties of each data source should be added in a request (for example **queries.stringInput** as shown in the request above). To better understand how to form a query for a certain data source, use the Developer Tools in your browser of choice and inspect the HTTP requests being made to `/api/ds/query`.
```

**Example Test data source time series query response:**

```json
{
  "results": {
    "A": {
      "frames": [
        {
          "schema": {
            "refId": "A",
            "fields": [
              {
                "name": "time",
                "type": "time",
                "typeInfo": {
                  "frame": "time.Time"
                }
              },
              {
                "name": "A-series",
                "type": "number",
                "typeInfo": {
                  "frame": "int64",
                  "nullable": true
                }
              }
            ]
          },
          "data": {
            "values": [
              [1644488152084, 1644488212084, 1644488272084, 1644488332084, 1644488392084, 1644488452084],
              [1, 20, 90, 30, 5, 0]
            ]
          }
        }
      ]
    }
  }
}
```

## Deprecated resources

> Please note that these resources have been deprecated and will be removed in a future release.

Reviewer: ```suggestion
The following resources have been deprecated. They will be removed in a future release.
```

  }
}
```

## Query a data source by ID

Queries a data source having backend implementation.

`POST /api/ds/query`

> **Note:** Most of Grafana's builtin data sources have backend implementation.

**Example request for the Test data source**:

```http
POST /api/ds/query HTTP/1.1
Accept: application/json
Content-Type: application/json

{
   "queries":[
      {
         "refId":"A",
         "scenarioId":"csv_metric_values",
         "datasource":{
            "uid":"PD8C576611E62080A",
            "type":"testdata"
         },
         "stringInput":"1,20,90,30,5,0",

>          "datasourceId":6,

Reviewer 2: Is this really here? And/or should we render it here now when we have the datasource ref above?

Author: Yep that's here alright. It's a legacy field AFAICT that gets auto set. Will try dig a bit further on it and see if it can be removed without breaking other things

Reviewer 2: As discussed, since datasource ref is included above we can leave this out in the example since we want users to migrate away from using `datasourceId`.

         "maxDataPoints":1848,
         "intervalMs":200,
      }
   ],
   "from":"now-5m",
   "to":"now"
}
```

> **Note:** The `from`, `to`, and `queries` properties are required.

JSON Body schema:

- **queries** – Specifies one or more queries. Must contain at least 1.
- **queries.refId** – Specifies an identifier of the query. Is optional and default to "A".
- **queries.datasource.uid** – Specifies the UID of data source to be queried. Each `query` in the request must have an unique `datasource`.
- **queries.datasource.type** – Specifies the data source plugin identifier to be queried.
- **queries.maxDataPoints** - Species maximum amount of data points that dashboard panel can render. Is optional and defaults to 100.
- **queries.intervalMs** - Specifies the time interval in milliseconds of time series. Is optional and defaults to 1000.
- **from/to** – Should be either absolute in epoch timestamps in milliseconds or relative using Grafana time units. For example, `now-5m`.

>   In addition, each data source has its own specific properties that should be added in a request.

Reviewer 2: Wonder if we can describe this in a way that's still helpful for users compared with how we've documented /api/tsdb/query that included some other examples, e.g. mysql. Maybe we can refer to something a long the lines "Please use the Developer Tools of your browser of choice and inspect the HTTP requests being made to `/api/ds/query` to understand how to query a certain datasource in question."?

Not sure if these spaces was intentional?

```suggestion
In addition, each data source has its own specific properties that should be added in a request.
```
